pkg/controllers/mcs/service_import_controller.go

```diff
@@ -26,6 +36,167 @@ const ServiceImportControllerName = "service-import-controller"
+		}
+	}
+
+	return nil
+}
+
+func (c *ServiceImportController) getServiceImportPropagateClusterNames(derivedSvc *corev1.Service) ([]string, error) {
+	clusterNames, clusterErr := c.getSvcClusterNameFromControlSvc(derivedSvc)
+	if clusterErr != nil {
+		return nil, clusterErr
+	}
+	return clusterNames, nil
+}
+
+func (c *ServiceImportController) getClusterIPFromClusterDerivedSvs(derivedSvc *corev1.Service, clusterDynamicClient *util.DynamicClusterClient) (string, error) {
+	dynamicResource, err := restmapper.GetGroupVersionResource(c.RESTMapper, schema.FromAPIVersionAndKind(derivedSvc.APIVersion, derivedSvc.Kind))
+	if err != nil {
+		return "", err
+	}
+
+	var svcLoad *unstructured.Unstructured
+	var getErr error
+	if waitErr := wait.PollUntilContextTimeout(context.TODO(), 1*time.Second, 10*time.Second, true, func(ctx context.Context) (bool, error) {
+		svcLoad, getErr = clusterDynamicClient.DynamicClientSet.Resource(dynamicResource).Namespace(derivedSvc.Namespace).Get(context.TODO(), derivedSvc.Name, metav1.GetOptions{})
+		if getErr != nil {
+			klog.Warningf("Failed to get derived service %s/%s from cluster, err: %v", derivedSvc.Namespace, derivedSvc.Name, err)
+			return false, nil
+		}
+		return true, nil
+	}); waitErr != nil {
+		klog.Errorf("Failed to get derived service %s/%s from cluster, err: %v", derivedSvc.Namespace, derivedSvc.Name, err)
+		return "", waitErr
+	}
+
+	clusterIP, _, err := unstructured.NestedString(svcLoad.Object, "spec", "clusterIP")
+	if err != nil {
+		return "", err
+	}
+	return clusterIP, nil
+}
+
+func (c *ServiceImportController) updateClusterServiceImport(svcImport *mcsv1alpha1.ServiceImport, clusterIP string, clusterDynamicClient *util.DynamicClusterClient) error {
+	dynamicResource, err := restmapper.GetGroupVersionResource(c.RESTMapper, schema.FromAPIVersionAndKind(svcImport.APIVersion, svcImport.Kind))
+	if err != nil {
+		return err
+	}
+
+	var clusterSvcImportObj *unstructured.Unstructured
+	var setErr error
+	if err := wait.PollUntilContextTimeout(context.TODO(), 1*time.Second, 10*time.Second, true, func(ctx context.Context) (bool, error) {
+		if clusterSvcImportObj, setErr = clusterDynamicClient.DynamicClientSet.Resource(dynamicResource).Namespace(svcImport.Namespace).Get(context.TODO(),
+			svcImport.Name, metav1.GetOptions{}); setErr != nil {
+			klog.Warningf("Failed to get service import %s/%s from cluster, err: %v", svcImport.Namespace, svcImport.Name, err)
+			return false, nil
+		}
+		return true, nil
+	}); err != nil {
+		klog.Errorf("Failed to get service import %s/%s from cluster, err: %v", svcImport.Namespace, svcImport.Name, err)
+		return err
+	}
+
+	unstructuredObj, addErr := c.addUnstructSvcImportClusterIP(clusterSvcImportObj, clusterIP)
+	if addErr != nil {
+		return addErr
+	}
+
+	if _, err = clusterDynamicClient.DynamicClientSet.Resource(dynamicResource).Namespace(unstructuredObj.GetNamespace()).Update(context.TODO(), unstructuredObj, metav1.UpdateOptions{}); err != nil {
```

Comment:
we may can `return clusterDynamicClient.DynamicClientSet.Resource(dynamicResource).Namespace(unstructuredObj.GetNamespace()).Update(context.TODO(), unstructuredObj, metav1.UpdateOptions{})` 🫢

Reply:
> we may can `return clusterDynamicClient.DynamicClientSet.Resource(dynamicResource).Namespace(unstructuredObj.GetNamespace()).Update(context.TODO(), unstructuredObj, metav1.UpdateOptions{})` 🫢

Yes, I've thought about that too.But the update function returns two values.
see here:

```go
Update(ctx context.Context, obj *unstructured.Unstructured, options metav1.UpdateOptions, subresources ...string) (*unstructured.Unstructured, error)
```

Reply:
oh sry, my bad